_posts/2020-01-21-flyway-spring-boot-multitenancy.md

```diff
@@ -0,0 +1,242 @@
+
+To keep it simple, let's consider the last option with sending `tenantId` as a header of the HTTP request.
+To read the header from the request we implement `WebRequestInterceptor` interface. This interface defines 
+an interceptor that is executed before the request is received in the `Controller`.
+
+```java
+@Component
+public class HeaderTenantInterceptor implements WebRequestInterceptor {
+
+    public static final String TENANT_HEADER = "X-tenant";
+
+    @Override
+    public void preHandle(WebRequest request) throws Exception {
+        ThreadTenantStorage.setTenantName(request.getHeader(TENANT_HEADER));
+    }
+}
+```
+Every incoming request is proceeded in the method `preHandle(WebRequest request)`. The tenant from the header is set
+to `ThreadTenantStorage`. `ThreadTenantStorage` is a storage, that contains a `ThreadLocal` variable. With `ThreadLocal` we can be sure, that
+every thread has its copy of this variable and the current thread has no access to a `ThreadLocal`
+of another tenant.
+
+```java
+public class ThreadTenantStorage {
+
+    private static ThreadLocal<String> currentTenant = new ThreadLocal<>();
+
+    public static void setTenantName(String tenantName) {
+        currentTenant.set(tenantName);
+    }
+
+    public static String getTenantName() {
+        return currentTenant.get();
+    }
+
+    public static void clear(){
+        currentTenant.remove();
+    }
+}
+```
+
+The last step in the configuration of tenant binding is to register the interceptor in the `WebMvcConfigurer`
+
+```java
+@Configuration
+public class WebConfiguration implements WebMvcConfigurer {
+
+    private final HeaderTenantInterceptor headerTenantInterceptor;
+
+    public WebConfiguration(HeaderTenantInterceptor headerTenantInterceptor) {
+        this.headerTenantInterceptor = headerTenantInterceptor;
+    }
+
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addWebRequestInterceptor(headerTenantInterceptor);
+    }
+}
+```
+
+## Multiple Data Source Configuration
+
+Normally we define the `DataSource` configuration in `application.yaml` using properties `spring.datasource`.
+But we can define only one `DataSource` with these properties. To load multiple `DataSource`s we
+can customize properties. 
+
+```yaml
+spring:
+  datasources:
+    vw:
+      jdbcUrl: jdbc:h2:mem:vw
+      driverClassName: org.h2.Driver
+      username: sa
+      password: password
+    bmw:
+      jdbcUrl: jdbc:h2:mem:bmw
+      driverClassName: org.h2.Driver
+      username: sa
+      password: password
+```
+In this case, two data sources are configured: `vw` and `bmw`. If we assume, that we had a fixed number of tenants
+we could define the `Datasource`s for them as static data. But we want to be able to add new tenants without building
+the application again. So let's implement it dynamically.
+Now we can bind the properties to a Spring Bean using `@ConfigurationProperties`.
```

Comment:
Ich verstehe das mit dem dynamischen implementieren nicht. Deine Loesung ist nicht dymanisch, oder verstehe ich das falsch?

Reply:
Ich habe einen Satz hinzugefügt, um es ausführlicher zu erklären.

Mit "dynamisch" meinte ich, dass die Tenants nicht im Code fest definiert sind, sondern in Konfiguration. Wenn man die Konfiguration externalisiert, dann kann man einen neuen Tenant hinzufügen ohne den Code neu zu kompilieren. Man müsste dann nur die Applikation neu starten um die Flyway Migration für alle Tenants anzutriggern. Würde ein anders Wort als "dynamisch" besser passen?

Reply:
Eine `application.yml` Datei kann man auch als externe Konfiguration benutzen, wenn man sie nicht mit ins JAR packt. Von daher ist der Absatz vermutlich einfach ueberfluessig. Ich hab ihn mal entfernt.